js/src/react/adapters.js

```diff
@@ -18,29 +18,44 @@ ShinyProxy.addCustomMessageHandler('updateReactInput', ({ inputId, data }) => {
   } else throw new Error(`Attempted to update non-existent React input '${inputId}'`);
 });
 
-export function InputAdapter(Component, valueProps) {
-  function Adapter({ inputId, value: defaultValue, ...otherProps }) {
-    const [value, setValue] = useState(defaultValue);
-    const [updatedProps, setUpdatedProps] = useState();
+function registerUpdateHandler(inputId, setValue, updatedProps, setUpdatedProps) {
+  const updateHandler = (props) => {
+    const { value, ...newProps } = props;
+    if (value !== undefined) setValue(value);
+    setUpdatedProps({ ...updatedProps, ...newProps });
+  };
+  updateHandlers[inputId] = updateHandler;
+  return () => {
+    // When the component is rerendered inside `renderUI()`, the new instance is initialised
+    // (and registers its update handler) *before* the old one is cleaned up. Here we ensure
+    // that the old instance only removes its own update handler.
+    if (updateHandlers[inputId] === updateHandler) delete updateHandlers[inputId];
+  };
+}
+
+function useValue(inputId, defaultValue) {
+  const [value, setValue] = useState(defaultValue);
+  useEffect(() => {
+    ShinyProxy.setInputValue(inputId, value);
+  }, [inputId, value]);
 
-    useEffect(() => {
-      ShinyProxy.setInputValue(inputId, value);
-    }, [inputId, value]);
+  return [value, setValue];
+}
 
-    // Register / cleanup update handlers (used to implement updateX.shinyInput functions).
-    useEffect(() => {
-      const updateHandler = ({ value: newValue, ...newProps }) => {
-        if (newValue !== undefined) setValue(newValue);
-        setUpdatedProps({ ...updatedProps, ...newProps });
-      };
-      updateHandlers[inputId] = updateHandler;
-      return () => {
-        // When the component is rerendered inside `renderUI()`, the new instance is initialised
-        // (and registers its update handler) *before* the old one is cleaned up. Here we ensure
-        // that the old instance only removes its own update handler.
-        if (updateHandlers[inputId] === updateHandler) delete updateHandlers[inputId];
-      };
-    }, [inputId]);
+function useUpdatedProps(inputId, setValue) {
+  const [updatedProps, setUpdatedProps] = useState();
+  useEffect(() => {
+    const handler = registerUpdateHandler(inputId, setValue, updatedProps, setUpdatedProps);
+    return handler;
```

Comment:
This gives the impression, that the function returns an update handler, but it actually returns a cleanup function for the effect :wink: I think it will be more readable to just inline this function here.